### PR TITLE
Block Bindings API: Add block bindings PHP registration mechanisms and "Post meta" source under the experimental flag

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
Related issues:

https://github.com/WordPress/gutenberg/issues/54536
https://github.com/WordPress/gutenberg/issues/53300
https://github.com/WordPress/gutenberg/pull/56867

I'm splitting [the block bindings API prototype](https://github.com/WordPress/gutenberg/pull/56867/) into smaller PRs so they are more manageable and discussions can be kept separately. You can find more information in the original PR.

This pull request covers:
* Change the "Test connections" experiment name to "Test block bindings & custom fields" and modified the code depending on that.
* Remove the old UI to add connections.
* Add the PHP mechanisms needed for the block bindings API:
  * A new PHP function `register_block_bindings_source` that allow users to create new sources, passing a name, a label and a callback.
  * Logic to replace the proper HTML depending on the block attribute source. Right now it is hardcoded for the allowed blocks (paragraph, heading, button, and image), but we could use the `set_inner_html` function from the HTML API once that is ready.
* Register the `post_meta` source, that binds block attributes and custom fields.
* Register the `pattern_attributes` source, that is needed for partially synced patterns.
* Adapt the partially synced patterns experiment to use the new system.
 
This pull request doesn't add a UI to create the bindings, so the attribute has to be added manually.

## Why?
This pull request is the basis for the block bindings API which will enable the connection between block attributes and other sources like custom fields.

As mentioned, I am splitting the original prototype into smaller PRs.

## Testing Instructions

For all the testing we have to go to Gutenberg-Experiments and enable the Test Block Bindings option.

**Test that adding the bindings attribute works**
1. Register a new custom field however you prefer. You can use a snippet similar to this:

```PHP
register_meta(
	'post',
	'text_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'This is the content of the text custom field',
	)
);
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

2. Create a new post/page and insert a paragraph.
3. Open the code editor view and modify the attributes of the paragraph to create the binding:
```
{"metadata":{"bindings":{"content":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}}}}
```
4. Go to the frontend and check that, independently of the content of the paragraph in the editor, it is changed by the custom field value in the server.

**Test that multiple attributes values can be connected**

Repeat the same process but using a button (or an image) to bind the content and the URL.

1. Add bindings for both attributes with something like this:

```
{"metadata":{"bindings":{"text":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}, "url":{"source":{"name":"post_meta","attributes":{"value":"url_custom_field"}}}}}}
```
2. Go to the frontend and check that both the text and the URL point to the custom field.

**Check that the old UI is not there**

In the existing experiment, the UI used doesn't make sense anymore.
1. Go to any paragraph.
2. Check that in the block panel in the right sidebar the Connections section doesn't exist anymore.

**Test that new sources can be registered**
1. Using a PHP snippet, or any method preferred, register a new source using the `register_block_bindings_source` function. In this case, I'm adding a "Shortcode" source and create a custom one to test it:
```
// // Register the shortcode source
if (function_exists('register_block_bindings_source')) {
    $shortcode_source_callback = function ( $source_attrs, $block_content, $block, $block_instance ) {
		return do_shortcode( $source_attrs['value'] );
	};
	register_block_bindings_source( 'shortcode', $shortcode_source_callback );
	register_block_bindings_source(
		'shortcode',
		array(
			'label' => __( 'Shortcode' ),
			'apply' => $shortcode_source_callback,
		)
	);
}

// Add a shortcode for testing
function my_custom_shortcode_callback() {
    return "This is my custom shortcode content.";
}
add_shortcode('my_custom_shortcode', 'my_custom_shortcode_callback');
```

2. In any paragraph, add a new attribute pointing to the new source you created:

```
{"metadata":{"bindings":{"content":{"source":{"name":"shortcode","attributes":{"value":"[my_custom_shortcode]"}}}}}}
```

3. Check in the frontend that the value of your source is shown.

**Test that partially synced patterns keep working as expected**
Partially synced patterns experiment should keep working as it does before this PR.

1. Go to the Site Editor -> Patterns -> Create a new synced pattern.
2. Add a heading and a paragraph.
3. In the block settings of the paragraph -> Advanced -> Toggle Synced Attribute "content" to enable editing the paragraph.
4. Insert that pattern into a post/page multiple times and define different values for the paragraphs.
5. Check in the frontend that the value persists.
6. Modify something in the synced pattern (layout for example) and check that is applied to the post/page you created.

### CODE REVIEW

**Review by artemiomorales — COMMENTED** [review]
Fantastic job with this and the related PRs :raised_hands: I’ve followed the testing instructions and everything works as expected. While I don’t have enough experience with Gutenberg to advise on best practices there, I have a few general thoughts.

At a high level, as per our conversation @SantosGuillamot, I think a minimal version like this without any visible UI could be good to merge into core for 6.5 (though perhaps a bit early for the partially synced patterns). Regardless, I think we can just focus on merging the prototype to Gutenberg and collecting feedback rather than worrying too much about core for the moment.

A heads up that both @c4rl0sbr4v0 and I found the customizable attributes in the partially synced patterns to have a confusing label, since these attributes are in fact the ones that will _not_ be synchronized. Would we be able to rename this heading from Synced Attributes to Customizable Attributes or Decoupled Attributes or another more apt name?

Screenshot 2023-12-22 at 3 28 43 PM

> Test that multiple attributes values can be connected
> Repeat the same process but using a button (or an image) to bind the content and the URL.

To help folks who are testing and may not have a chance to look in-depth at the code, we should note that currently 1.) for the button, only the `url` and `text` attributes work, and 2.) for the image, only the `url` and `title` attributes work. On that note, how about we also expose the `alt` attribute for the image? I know we’re mostly in a phase of getting feedback at the moment, but that small change would make this more useable in the wild.

Lastly, I saw in the [previous PR's description](https://github.com/WordPress/gutenberg/pull/56867/) that you were still undecided on the format of the bindings property. Has that discussion now been settled? Note that I don’t have a strong opinion in any direction — I’m just curious to hear if it's been resolved or requires further input.

Will continue looking at the code and offer additional thoughts as they arise. Thanks :pray:

**Comment by SantosGuillamot** [comment]
Thanks for sharing your thoughts, Artemio 🙂 

> Would we be able to rename this heading from Synced Attributes to Customizable Attributes or Decoupled Attributes or another more apt name?

Would you mind sharing this feedback in the Synced Patterns issue: [link](https://github.com/WordPress/gutenberg/issues/53705)? This PR just focuses on the block bindings API, that is used by synced patterns. But their UI is independent of this.

> On that note, how about we also expose the alt attribute for the image?

`alt` attribute should work exactly like the `title` so there should be no problem adding that. I just added a few attributes to understand the system better, but we can definitely add `alt` as well.

> Lastly, I saw in the https://github.com/WordPress/gutenberg/pull/56867/ that you were still undecided on the format of the bindings property. Has that discussion now been settled?

I discussed it with some folks, and we agreed on an initial syntax, trying to leave as few remaining decisions as possible. But if you, or anyone else, believes another syntax is better, we can still modify it.

**Comment by artemiomorales** [comment]
> Would you mind sharing this feedback in the Synced Patterns issue: https://github.com/WordPress/gutenberg/issues/53705? This PR just focuses on the block bindings API, that is used by synced patterns. But their UI is independent of this.

I just checked out `trunk` and it appears the wording was changed to `Pattern Overrides: Allow instance overrides` in the following PR, so all good on this front: https://github.com/WordPress/gutenberg/pull/57009

Screenshot 2024-01-02 at 2 30 48 PM  

> I discussed it with some folks, and we agreed on an initial syntax, trying to leave as few remaining decisions as possible. But if you, or anyone else, believes another syntax is better, we can still modify it.

Ok great. I think the initial syntax we have is good 👍

File: packages/patterns/src/components/partial-syncing-controls.js
Comment (artemiomorales) [inline packages/patterns/src/components/partial-syncing-controls.js:54]: Should we remove the `metadata` attribute if it's no longer in use? Otherwise it seems like the pattern will just contain unnecessary information indefinitely, though I could be missing something.

File: packages/patterns/src/components/partial-syncing-controls.js
Comment (artemiomorales) [inline packages/patterns/src/components/partial-syncing-controls.js:19]: Not sure if this PR would be the best place to do this, but I think we should rename this to `editableAttributes` or something similar for clarity, as these are the attributes that users will be able to override.

**Comment by artemiomorales** [comment]
I renamed some variables for clarity and cleaned up some code around the button markup in https://github.com/WordPress/gutenberg/pull/57249/commits/176da0abd96cdf6240bdfb85fea61b4f82194243, though feel free to change it back or continue modifying.

As far as I can tell, everything in the PR still seems to work.

**Comment by kevin940726** [comment]
> A heads up that both @c4rl0sbr4v0 and I found the customizable attributes in the partially synced patterns to have a confusing label, since these attributes are in fact the ones that will _not_ be synchronized. Would we be able to rename this heading from Synced Attributes to Customizable Attributes or Decoupled Attributes or another more apt name?

We recently modified the control to be a single checkbox in https://github.com/WordPress/gutenberg/pull/57009. We just need a rebase here 🙂. I'm happy to help too!

File: packages/patterns/src/components/partial-syncing-controls.js
Comment (kevin940726) [reply packages/patterns/src/components/partial-syncing-controls.js:54]: One possible use case would be retaining the `id` when the user is just ticking the checkbox on and off. We wouldn't want to regenerate an id every time in this case.

File: lib/experimental/block-bindings/sources/pattern.php
Comment (kevin940726) [inline lib/experimental/block-bindings/sources/pattern.php]: Should we return `null` here?

File: lib/experimental/block-bindings/sources/pattern.php
Comment (kevin940726) [inline lib/experimental/block-bindings/sources/pattern.php]: Same here. We should return `null` if there's no overrides for this block or else the frontend will just render an empty content.

```suggestion
		return _wp_array_get( $block_instance->context, array( 'pattern/overrides', $block_id ), null );
```

**Comment by artemiomorales** [comment]
> We recently modified the control to be a single checkbox in https://github.com/WordPress/gutenberg/pull/57009. We just need a rebase here 🙂. I'm happy to help too!

@kevin940726 Great, thanks! I've opened a [throwaway PR](https://github.com/WordPress/gutenberg/pull/57558) in my fork to illustrate how I'm thinking to approach the rebase. I believe it works as expected, but would you be able to take a look?

Note that I [added a commit](https://github.com/WordPress/gutenberg/pull/57558) to attempt to actually use the multiple overridden attributes in the pattern, but it did not work for the image — only one attribute was read and it was erroneously used for both the `title` and `alt`.

I see [here](https://github.com/WordPress/gutenberg/pull/57558/commits/2d5d7805fc2c2944125036dfc6c0936b9f73806d) however that support for multiple attributes does not yet seem to be added in the editor code for the partially synced pattern. Am I following this correctly?

I just want to check in before I get too deep. Any guidance appreciated!

**Comment by kevin940726** [comment]
> Note that I [added a commit](https://github.com/WordPress/gutenberg/pull/57558) to attempt to actually use the multiple overridden attributes in the pattern, but it did not work for the image — only one attribute was read and it was erroneously used for both the `title` and `alt`.
> 
> I see [here](https://github.com/WordPress/gutenberg/pull/57558/commits/2d5d7805fc2c2944125036dfc6c0936b9f73806d) however that support for multiple attributes does not yet seem to be added in the editor code for the partially synced pattern. Am I following this correctly?

Yes, that's correct! We don't currently support them at the time, but we're planning to add support soon.

The rebase PR looks good! It's mostly the same as what I did 😆. I think the next step for us would be adding support for `core/heading` and `core/button`. Then, we'll consider adding support for `core/image` (and multiple attributes).

**Comment by artemiomorales** [comment]
Noting here that I've come across a bug — when enabling pattern overrides, the original content disappears when the pattern is used in an instance.

I won't have time to look at this until next week, but I'm documenting it here in case anyone is able to look.

https://github.com/WordPress/gutenberg/assets/5360536/aad67dba-15cf-4ec9-87c6-95c99b743155

**Comment by artemiomorales** [comment]
> The rebase PR looks good! It's mostly the same as what I did 😆. I think the next step for us would be adding support for core/heading and core/button. Then, we'll consider adding support for core/image (and multiple attributes).

Ok perfect. I've rebased on `trunk` and force pushed the branch. Just in case anyone needs it, I created [a backup](https://github.com/artemiomorales/gutenberg/tree/add/block-bindings-php-registration-mechanisms-backup) with the original commit history on my fork.

I see that a [PR has been opened](https://github.com/WordPress/gutenberg/pull/57573) to support the multiple attributes — thanks!

@gziolo has advised that the block bindings and partially synced patterns experiments should both work seamlessly for this prototype, so we should [fix the outstanding bug](https://github.com/WordPress/gutenberg/pull/57249#issuecomment-1878673920) and sync with the new PR before merging.

**Comment by kevin940726** [comment]
> Noting here that I've come across a bug — when enabling pattern overrides, the original content disappears when the pattern is used in an instance.

I think this is the same bug I mentioned in https://github.com/WordPress/gutenberg/pull/57249#discussion_r1441287871.

File: lib/experimental/block-bindings/sources/index.php
Comment (talldan) [inline lib/experimental/block-bindings/sources/index.php:9]: A lot of the registries in WordPress use the approach of having a singleton class, so it could be something to replicate here.

Some examples:
- Block Pattern Categories: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-block-pattern-categories-registry.php
- Block Styles: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-block-styles-registry.php

If a class is used, it's worth thinking about how it might be extended by the gutenberg codebase when shipped in core if changes need to be made to block bindings.

File: lib/experimental/blocks.php
Comment (talldan) [inline lib/experimental/blocks.php:99]: Regarding the comment, I don't think it's worth considering an opt-in system right now. It's a little unclear how it would work. While the number of supported blocks is quite small, it seems fine as private internal config.

It'd be good to avoid a global variable, since that opens this up to being used by third parties, essentially making it a public API.

File: packages/block-library/src/paragraph/block.json
Comment (talldan) [inline packages/block-library/src/paragraph/block.json]: Right now there's no such thing as an experimental block supports, this will be considered stable, so I think it'd be best to either:
a) Remove the `__experimental` prefix
b) Avoid using the block supports system until the bindings API is more stable

There's also some misalignment with how a config object is used server side and this supports setting is used client side. I'd vote for using config everywhere until the number of supported blocks is such a small subset.

**Review by talldan — COMMENTED** [review]
This is looking good so far, thanks for breaking the initial PR up into smaller parts.

File: lib/experimental/block-bindings/sources/index.php
Comment (talldan) [inline lib/experimental/block-bindings/sources/index.php]:
```suggestion
	 * @param array  $source_args List of arguments for the block bindings source:
```

File: lib/experimental/block-bindings/sources/index.php
Comment (talldan) [inline lib/experimental/block-bindings/sources/index.php]: The description for `apply` could be improved a bit, as this is probably the most important part of the API, and also the hardest for API users to understand.

Some things they'll need to know:
- When is this called?
- What are the params?
- What should be returned?

File: lib/experimental/block-bindings/sources/pattern.php
Comment (talldan) [inline lib/experimental/block-bindings/sources/pattern.php]: The params for `apply` are probably the one thing in this PR that need to be carefully considered for the future (in terms of the back compat problems it might introduce). 

- `$source_attrs` - I see this is a value taken from the attributes of the block being processed, so it is a little bit of a duplication given the block instance is already provided. It is a handy convenience.
- `$block_content` - It seems to be unusued, and I'm wondering what potential future usage there might be for this? I think it's also already available via `$block_instance->inner_html`, so perhaps not necessary as a param.
- `$block` / `$block_instance` - `$block` represents the parsed block object, which is also available using `$block_instance->parsed_block`, so I think it could be removed as a param.

I think it could be reduced to two params, possibly even one.

File: lib/experimental/block-bindings/sources/pattern.php
Comment (artemiomorales) [reply lib/experimental/block-bindings/sources/pattern.php]: Done, thanks! 🙌

File: lib/experimental/blocks.php
Comment (SantosGuillamot) [reply lib/experimental/blocks.php:99]: I agree we probably won't need an opt-in system right now, although it's something we'll need to consider in the future.

> It'd be good to avoid a global variable, since that opens this up to being used by third parties, essentially making it a public API.

Makes sense. We can remove it for this initial phase and figure out later how to do the opt-in system.

File: packages/block-library/src/paragraph/block.json
Comment (SantosGuillamot) [reply packages/block-library/src/paragraph/block.json]: If I remember correctly, I just kept that attribute because the partially synced patterns were using it. From [the second pull request](https://github.com/WordPress/gutenberg/pull/57258), it shouldn't be needed for anything else.

It'd be good to double check that anyway.

File: lib/experimental/block-bindings/sources/index.php
Comment (SantosGuillamot) [reply lib/experimental/block-bindings/sources/index.php:9]: Thanks for sharing! @michalczaplinski @artemiomorales It'd be great if you can explore this possibility.

File: lib/experimental/block-bindings/sources/index.php
Comment (SantosGuillamot) [reply lib/experimental/block-bindings/sources/index.php]: Makes sense 🙂  Let's improve this part as well.

File: lib/experimental/block-bindings/sources/pattern.php
Comment (SantosGuillamot) [reply lib/experimental/block-bindings/sources/pattern.php]: Regarding this, I passed those params because we are using the [render_block_{$this->name} hook](https://developer.wordpress.org/reference/hooks/render_block_this-name/) to process the bindings: [link](https://github.com/WordPress/gutenberg/pull/57249/files#diff-be8a354ce9c01e85c0fdf04c75d76880cdc6c81992db7a86a7a0515dc11a9833R169). As that filter is using those params, I decided to pass them in case they are needed and other sources want to use them.

Having said that, I agree it's worth analyzing if they are really needed and try to simplify the `apply`. To ensure we are aligned, you propose using `block_instance` for everything, right? For example, this would mean that the pattern and the post meta sources would be:

**Pattern**
```PHP
// Before
$pattern_source_callback = function ( $source_attrs, $block_content, $block, $block_instance ) {
   ...
}

// After
$pattern_source_callback = function ( $block_instance ) {
   ...
}
```

**Post meta**

```PHP
// Before
$post_meta_source_callback = function ( $source_attrs ) {
   ...
}

// After
$post_meta_source_callback = function ( $block_instance ) {
   $source_attrs = $block_instance->parsed_block['attrs']['metadata']['bindings'];
   ...
}
```

Is that what you had in mind?

For the sources we have created so far in the different experiments, in all of them we are just using the `source_attrs` aside from the `pattern` where we are using the `$block_instance`.

File: lib/experimental/block-bindings/sources/index.php
Comment (artemiomorales) [reply lib/experimental/block-bindings/sources/index.php:9]: Thanks! Agreed that we should begin improving the code as we look to bring this into Core.

In order to keep things moving and start getting more feedback on the feature, and since this is still an experiment, I think we can get this PR merged first then immediately begin another PR to refactor and start moving the logic over to a Block Bindings class 👍

File: packages/block-library/src/paragraph/block.json
Comment (gziolo) [reply packages/block-library/src/paragraph/block.json]: As long as all blocks and attributes allowed are locked in the code, we can omit block supports from the initial scope. We will definitely need it later. Maybe, it’s going to be needed for V1 as a way for extenders to opt out from the feature. It’s a separate discussion to have that shouldn’t block the progress on the prototype.

File: lib/experimental/block-bindings/sources/index.php
Comment (gziolo) [reply lib/experimental/block-bindings/sources/index.php]: It seems like a low hanging fruit so let’s address before landing this PR.

File: lib/experimental/block-bindings/sources/pattern.php
Comment (gziolo) [reply lib/experimental/block-bindings/sources/pattern.php]: My impression is that we should refactor the way bindings get registered first, and it will drive changes in how these filters work. In the WordPress core, we won’t use filters as far as I can tell from a  quick assessment so that might evolve into a bit different shape.

File: lib/experimental/block-bindings/sources/index.php
Comment (gziolo) [reply lib/experimental/block-bindings/sources/index.php:9]: Great feedback, @talldan. That’s the direction I would recommend, too. I  agree with @artemiomorales as well that this refactoring might be easier to perform as a focused effort in a follow up PR. All functionality is behind an experiment. Well, another option is to open a new branch targeting this branch if you decide to spend more time polishing this PR.

Some inspiration for how to architect the class and helper functions can be borrowed from the recent work on ES Modules API in https://github.com/WordPress/wordpress-develop/pull/5818 where @felixarntz did an outstanding job explaining several good practices to take into account. That combined with prior work referenced by Dan should give you necessary insights to refactor the code.

File: lib/experimental/block-bindings/sources/post-meta.php
Comment (michalczaplinski) [inline lib/experimental/block-bindings/sources/post-meta.php]: On line 10 it says "otherwise use context" but here we mention that context
wasn't available. Should we update that first comment?

File: packages/block-library/src/paragraph/block.json
Comment (artemiomorales) [reply packages/block-library/src/paragraph/block.json]: Ok I removed this and also [removed the block supports requirement](https://github.com/WordPress/gutenberg/pull/57249/commits/7fb646cae16f1703cfdee25a7f8629e6b020fec3) for the partially synced patterns.

File: lib/experimental/block-bindings/sources/pattern.php
Comment (artemiomorales) [reply lib/experimental/block-bindings/sources/pattern.php]: Ok I'll leave this as is for now and we can revise this as we being refactoring the code.

Comment (artemiomorales) [reply lib/experimental/block-bindings/sources/pattern.php]: I actually updated it to use just the values that are actually in use for now. We can change it back if we feel that will spark better discussion.

File: lib/experimental/block-bindings/sources/index.php
Comment (artemiomorales) [reply lib/experimental/block-bindings/sources/index.php]: Ok I updated the comments. I wasn't sure how to handle clarifying comments for code that isn't currently being used and/or is also temporary, so I updated the callback and did the best I could.

File: lib/block-supports/pattern.php
Comment (artemiomorales) [inline lib/block-supports/pattern.php:19]: ~~If we're removing block supports. we should probably remove this entire file, which exists inside the `block-supports/` directory, and put its logic elsewhere.~~ EDIT: Since we may be adding block supports in the future, perhaps removing this file isn't necessary.

File: packages/block-library/src/block/edit.js
Comment (artemiomorales) [inline packages/block-library/src/block/edit.js]: We should also remove this block supports.

**Review by artemiomorales — APPROVED** [review]
I'm approving with the caveat that there's still a lot to improve in terms of the API as well as removing the old block supports code and other improvements.

There's already a lot happening here though so I think it makes sense to begin addressing these in follow up PRs, as the partially synced patterns cleanup especially seems to be moving beyond the scope of what's being proposed here.

**Outstanding Items**
- ~~Finish removing block supports~~
- ~~Reformulate registration mechanisms~~
- Begin implementing singleton pattern via a Block Bindings class 
- ~~I noticed on `trunk` today that the block UI has gone missing from pattern instances when the partially synced patterns experiment is enabled, and while not caused by this PR, it prevents the functionality here from working and also needs to be fixed.~~ Edit: It turns out this is now [expected behavior](https://github.com/WordPress/gutenberg/issues/53705#issuecomment-1882305331).

Screenshot 2024-01-09 at 4 59 42 PM

**Review by michalczaplinski — COMMENTED** [review]
Agree with Artemio's review above. 

I would only add more context to this: 
> Reformulate registration mechanisms

I think that there are 2 aspects to this: 
1. Reconsidering what parameters should be passed to the `apply` as already mentioned by Dan in https://github.com/WordPress/gutenberg/pull/57249#discussion_r1444270854
2. Possibly simplifying the signature of `register_block_bindings_source`. 
  I believe that instead of `register_block_bindings_source( $source_name, $source_args )` where `$source_args is an associative array, the function could have 3 parameters: 
  ```php
  function register_block_bindings_source( $source_name, $label, $apply) {
    global $block_bindings_sources;
    $block_bindings_sources[ $source_name ] =  array( $label, $apply );
  }
  ```
  
  This way, the PHPDoc is much simpler and the signature is simplified as well.
  
I think that this and removing the block supports are the only 2 items we should address before merging.

**Comment by artemiomorales** [comment]
Ok @michalczaplinski I've [removed all of the block supports](https://github.com/WordPress/gutenberg/pull/57249/commits/5231b1cf0dedcd21f3792a4844613e5f94be6b95) and hardcoded support for just the paragraph block, as that's [all that's currently enabled](https://github.com/WordPress/gutenberg/blob/trunk/packages/patterns/src/constants.js) for pattern overrides.

I've also [updated the signature](https://github.com/WordPress/gutenberg/pull/57249/commits/5a212e6f5ec74a5b2f2feabd2f7cf24f8799b324) of the register function.

Let me know what you think!

**Review by kevin940726 — APPROVED** [review]
I think this is good to go as a first step. As per Pattern Overrides we can add a follow-up to support both `core/heading` and `core/button`. Then we can also work on adding support for `core/image` if possible!

**Review by michalczaplinski — APPROVED** [review]
Looks good to me! 

I've only done a minor adjustment of the PHPDoc string.

We can improve on it in the follow ups!

**Comment by gziolo** [comment]
@kevin940726, does it mean that Pattern Overrides work only with the Paragraph block at the moment?  Who is working on the follow-up PR?

**Comment by kevin940726** [comment]
> @kevin940726, does it mean that Pattern Overrides work only with the Paragraph block at the moment? Who is working on the follow-up PR?

Yes. Because both `core/heading` and `core/button` rely on this PR. It should be trivial to implement now that this is merged though!

File: lib/experimental/blocks.php
Comment (kevin940726) [inline lib/experimental/blocks.php:157]: I was just reminded that if the attribute accepts `null` or `undefined` as valid values for different meanings, then we should still be able to apply overrides here. Do you think we need a special value here to represent `none`? 🤔

Comment (artemiomorales) [reply lib/experimental/blocks.php:157]: @kevin940726 can you elaborate?
What do you think @SantosGuillamot @michalczaplinski?

I wonder how we would handle a `null` or `undefined` value as valid [here](https://github.com/WordPress/gutenberg/blob/a99575c2ea44bc5c5fd172e755f52c16b945d299/lib/experimental/block-bindings/class-wp-block-bindings.php#L139) or [here](https://github.com/WordPress/gutenberg/blob/a99575c2ea44bc5c5fd172e755f52c16b945d299/lib/experimental/block-bindings/class-wp-block-bindings.php#L104). It seems to me that what makes sense intuitively is for `null` or `undefined` to mean that there is no valid value to apply, but I also don't know enough about how developers would use this in the wild to have a strong opinion either way.

Comment (kevin940726) [reply lib/experimental/blocks.php:157]: For context, I was trying to add support for the `linkTarget` attribute of the button block. It accepts a string (like `_blank`) or `undefined` for no target. Let's say we have a source pattern with a button block and its `linkTarget` is `_blank`. Now we want to override the source pattern in an instance to remove the `linkTarget`. We have two problems here:

1. On the editor side, we need a way to represent `undefined` because it's not JSON serializable.
2. On the backend, we need a way to distinguish `NULL` from "skipping the attribute" because sometimes `NULL` is a valid value for some attributes.

I opened two draft PRs exploring different solutions for this problem and I'd love all your feedback!

- #57947: Treats non-existent attributes the same as `null` and uses `null` throughout to represent the attribute.
- #57993: Assumes `undefined` and `null` could be handled differently, and uses a patch similar to JSON patch to represent overrides. This also opens up future possibilities to represent "inserting" a partial override.

Note that the latter changes the `overrides` attribute format so it'd be better if we decide the format sooner than later to save us from back-compat pain 😅.

**Pull request merged**
